src/Data/Map/Internal.purs

```diff
@@ -737,3 +689,294 @@ mapMaybe = mapMaybeWithKey <<< const
+  Leaf ->
+    case r of
+      Leaf ->
+        Node 1 1 k v l r
+      Node h2 s2 _ _ _ _ ->
+        Node (1 + h2) (1 + s2) k v l r
+  Node h1 s1 _ _ _ _ ->
+    case r of
+      Leaf ->
+        Node (1 + h1) (1 + s1) k v l r
+      Node h2 s2 _ _ _ _ ->
+        Node (1 + if h1 > h2 then h1 else h2) (1 + s1 + s2) k v l r
+
+-- | Low-level Node constructor which maintains the balance invariants.
+-- | This is unsafe because it assumes the child Maps are ordered.
+unsafeBalancedNode :: forall k v. Fn4 k v (Map k v) (Map k v) (Map k v)
+unsafeBalancedNode = mkFn4 \k v l r -> case l of
+  Leaf ->
+    case r of
+      Leaf ->
+        singleton k v
+      Node rh _ rk rv rl rr
+        | rh > 1 ->
+            runFn7 rotateLeft k v l rk rv rl rr
+      _ ->
+        runFn4 unsafeNode k v l r
+  Node lh _ lk lv ll lr ->
+    case r of
+      Node rh _ rk rv rl rr
+        | rh > lh + 1 ->
+            runFn7 rotateLeft k v l rk rv rl rr
+        | lh > rh + 1 ->
+            runFn7 rotateRight k v lk lv ll lr r
+      Leaf
+        | lh > 1 ->
+            runFn7 rotateRight k v lk lv ll lr r
+      _ ->
+        runFn4 unsafeNode k v l r
+  where
+  rotateLeft :: Fn7 k v (Map k v) k v (Map k v) (Map k v) (Map k v)
+  rotateLeft = mkFn7 \k v l rk rv rl rr -> case rl of
+    Node lh _ lk lv ll lr
+      | lh > height rr ->
+          runFn4 unsafeNode lk lv (runFn4 unsafeNode k v l ll) (runFn4 unsafeNode rk rv lr rr)
+    _ ->
+      runFn4 unsafeNode rk rv (runFn4 unsafeNode k v l rl) rr
+
+  rotateRight :: Fn7 k v k v (Map k v) (Map k v) (Map k v) (Map k v)
+  rotateRight = mkFn7 \k v lk lv ll lr r -> case lr of
+    Node rh _ rk rv rl rr
+      | height ll <= rh ->
+          runFn4 unsafeNode rk rv (runFn4 unsafeNode lk lv ll rl) (runFn4 unsafeNode k v rr r)
+    _ ->
+      runFn4 unsafeNode lk lv ll (runFn4 unsafeNode k v lr r)
+
+  height :: Map k v -> Int
+  height = case _ of
+    Leaf -> 0
+    Node h _ _ _ _ _ -> h
+
+-- | Low-level Node constructor from two Maps.
+-- | This is unsafe because it assumes the child Maps are ordered.
+unsafeJoinNodes :: forall k v. Fn2 (Map k v) (Map k v) (Map k v)
+unsafeJoinNodes = mkFn2 case _, _ of
+  Leaf, b -> b
+  Node _ _ lk lv ll lr, r -> do
+    let (SplitLast k v l) = runFn4 unsafeSplitLast lk lv ll lr
+    runFn4 unsafeBalancedNode k v l r
+
+data SplitLast k v = SplitLast k v (Map k v)
+
+-- | Reassociates a node by moving the last node to the top.
+-- | This is unsafe because it assumes the key and child Maps are from
+-- | a balanced node.
+unsafeSplitLast :: forall k v. Fn4 k v (Map k v) (Map k v) (SplitLast k v)
+unsafeSplitLast = mkFn4 \k v l r -> case r of
+  Leaf -> SplitLast k v l
+  Node _ _ rk rv rl rr -> do
+    let (SplitLast k' v' t') = runFn4 unsafeSplitLast rk rv rl rr
+    SplitLast k' v' (runFn4 unsafeBalancedNode k v l t')
+
+data Split k v = Split (Maybe v) (Map k v) (Map k v)
+
+-- | Reassocates a Map so the given key is at the top.
```

Comment:
Shouldn't this comment say
```diff
-so the given key is at the top
+so the given key's value is at the top
```

Reply:
I don't think so. The shape of the tree is based on key ordering. It has nothing to do with values. It's just returning the value associated with that key so the node can be reassembled.